File: fathersDay/Dsa/Day 24 (345. Reverse Vowels of a String)/Solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_set>

class Solution {
public:
    std::string ReverseVowels(std::string s) {
        std::unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
        int left = 0, right = s.length() - 1;
        while (left < right) {
            while (left < right && vowels.find(s[left]) == vowels.end()) {
                left++;
            }
            while (left < right && vowels.find(s[right]) == vowels.end()) {
                right--;
            }
            if (left < right) {
                std::swap(s[left++], s[right--]);
            }
        }
        return s;
    }
};

int main() {
    Solution solution;
    std::string s = "hello";
    std::string result = solution.ReverseVowels(s);
    std::cout << '"' << result << '"' << std::endl; // Output: "holle"
    return 0;
}

File: fathersDay/Dsa/Day 24 (345. Reverse Vowels of a String)/Solution.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class Solution {
    public String ReverseVowels(String s) {
        Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));
        char[] sChars = s.toCharArray();
        int left = 0, right = sChars.length - 1;

        while (left < right) {
            while (left < right && !vowels.contains(sChars[left])) {
                left++;
            }
            while (left < right && !vowels.contains(sChars[right])) {
                right--;
            }
            if (left < right) {
                char temp = sChars[left];
                sChars[left++] = sChars[right];
                sChars[right--] = temp;
            }
        }
        return new String(sChars);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String s = "hello";
        String result = solution.ReverseVowels(s);
        System.out.println('"' + result + '"'); // Output: "holle"
    }
}
